p2p/gossip/protocol.go

```diff
@@ -2,405 +2,340 @@ package gossip
 
 import (
 	"errors"
-	"fmt"
-	"github.com/davecgh/go-spew/spew"
+	"github.com/golang/protobuf/proto"
 	"github.com/spacemeshos/go-spacemesh/crypto"
 	"github.com/spacemeshos/go-spacemesh/log"
 	"github.com/spacemeshos/go-spacemesh/p2p/config"
-	"github.com/spacemeshos/go-spacemesh/p2p/message"
-	"github.com/spacemeshos/go-spacemesh/p2p/net"
 	"github.com/spacemeshos/go-spacemesh/p2p/node"
+	"github.com/spacemeshos/go-spacemesh/p2p/pb"
+	"github.com/spacemeshos/go-spacemesh/p2p/service"
 	"hash/fnv"
 	"sync"
 	"time"
 )
 
-// NoResultsInterval is the timeout we wait between requesting more peers repeatedly
-const NoResultsInterval = 1 * time.Second
+const messageQBufferSize = 100
 
-// PeerMessageQueueSize is the len of the msgQ each peer holds.
-const PeerMessageQueueSize = 100
+const ProtocolName = "/p2p/1.0/gossip"
+const protocolVer = "0"
 
-// ConnectingTimeout is the timeout we wait when trying to connect a neighborhood
-const ConnectingTimeout = 20 * time.Second //todo: add to the config
+type hash uint32
 
 // fnv.New32 must be used everytime to be sure we get consistent results.
-func generateID(msg []byte) uint32 {
+func calcHash(msg []byte) hash {
 	msghash := fnv.New32() // todo: Add nonce to messages instead
 	msghash.Write(msg)
-	return msghash.Sum32()
+	return hash(msghash.Sum32())
 }
 
-// Protocol is a simple declaration of the gossip protocol
-type Protocol interface {
-	Broadcast(payload []byte) error
-
-	Start() error
-	Close()
-
-	Initial() <-chan struct{}
-	AddIncomingPeer(node.Node, net.Connection)
-	Disconnect(key crypto.PublicKey)
-}
-
-// PeerSampler is a our interface to select peers
-type PeerSampler interface {
-	SelectPeers(count int) []node.Node
+// Interface for the underlying p2p layer
+type BaseNetwork interface {
+	SendMessage(peerPubKey string, protocol string, payload []byte) error
+	RegisterProtocol(protocol string) chan service.Message
+	SubscribePeerEvents() (conn chan crypto.PublicKey, disc chan crypto.PublicKey)
+	ProcessProtocolMessage(sender node.Node, protocol string, data service.Data) error
 }
 
-// ConnectionFactory is our interface to get connections
-type ConnectionFactory interface {
-	GetConnection(address string, pk crypto.PublicKey) (net.Connection, error)
+type Signer interface {
```

Comment:
let's consider to not export this struct. though assuming both the other side knows the interface isn't so good aswell.